### PR TITLE
Curvy turn arrows

### PR DESCRIPTION
**Pull request by mdejean**
1. Use Bezier curves for all turns, so a lane change is now an offset bend instead of a straight line.
2. Change turn::angle to be the relative angle of the turn (i.e. +90 degrees is a normal left turn)
3. Draw turn marking arrows prettily using Bezier curves

### CODE REVIEW

File: map_model/src/objects/turn.rs [hunk 1]
```diff
@@ -110,7 +110,14 @@ impl Turn {
     // TODO What should this be for zero-length turns? Probably src's pt1 to dst's pt2 or
```
Comment (dabreegster) [inline map_model/src/objects/turn.rs:110]: Could you add a docstring here explaining this is a relative angle? (It really should've had something before, but I've only recently been better about adding docs...)

File: map_model/src/objects/turn.rs [hunk 2]
```diff
@@ -110,7 +110,14 @@ impl Turn {
...
     // something.
     pub fn angle(&self) -> Angle {
-        self.geom.first_pt().angle_to(self.geom.last_pt())
+        if self.geom.points().len() < 3 {
+            return Angle::ZERO;
+        }
+        //this is ridiculous, using segments of the turn geometry rather than the start and end angles
```
Comment (dabreegster) [inline map_model/src/objects/turn.rs:116]: Are you saying we'd get better results or it'd be more clear to use the origin and destination lane's geometry instead?

File: map_gui/src/render/lane.rs [hunk 3]
```diff
@@ -328,33 +326,119 @@ fn calculate_turn_markings(map: &Map, lane: &Lane) -> Vec<Polygon> {
...
-    // Don't call out the strange lane-changing in intersections. Per target road, find the average
-    // turn angle.
-    let mut angles_per_road: HashMap<RoadID, Vec<Angle>> = HashMap::new();
-    for turn in map.get_turns_from_lane(lane.id) {
-        angles_per_road
-            .entry(turn.id.dst.road)
-            .or_insert_with(Vec::new)
-            .push(turn.angle());
-    }
+    // Only show one arrow per road. They should all be the same angle, so just use last.
```
Comment (dabreegster) [inline map_gui/src/render/lane.rs:329]: An unexpected bit I noticed is
Screenshot from 2021-11-22 11-46-28
but that matches how things work before this PR. (It's just something we should reconsider -- it feels weird to be explicit about one lane's turns, but not another.)

**Review by dabreegster — COMMENTED** [review]
I tried this out and glanced through the code, and everything looks good. Huge improvement to visual quality! Anything else you wanted to do in this PR -- aka why is it a draft? And are there any edge cases you encountered still causing problems? I can run a screenshot diff test later to try and flush them out. (Or you can, in debug mode "screenshot all of the everything", then run `game/compare_screencaps.sh`. There are a few dependencies and extra input files you might need to do this though)

Example before:
Screenshot from 2021-11-22 11-45-16
after:
Screenshot from 2021-11-22 11-45-29

File: map_model/src/objects/turn.rs (hunk 2 above)
Comment (mdejean) [reply map_model/src/objects/turn.rs:116]: Since the turn geometry is converted from a bezier curve into 5 line segments the angle of the first and last line segment don't completely match the angle of the roads - that's why the turn arrows never seem to be quite turned enough. One possibility would be to store the angle in Turn, but that feels janky.

I'd like to think about not storing turn geometry at all and replacing the intersection in conflicts_with to something like a table lookup, then using curvy turns only for animation. That seems like a big change beyond the scope of this, though.

File: map_gui/src/render/lane.rs (hunk 3 above)
Comment (mdejean) [reply map_gui/src/render/lane.rs:329]: I agree that having arrows on only some lanes looks weird. Done by [line 311-327](https://github.com/a-b-street/abstreet/blob/5010acbc13ceecf5f06be8a6e8b6d5ea19341de0/map_gui/src/render/lane.rs#L311). That could instead be like if every lane on this road connects to every other possible outbound lane of the same type to provide the behavior you want.

**Comment by mdejean** [comment]
> And are there any edge cases you encountered still causing problems?

My main concern for edge cases is with the part that makes all turn geometry with curves and chooses the control points. There are many different shapes for intersections and I don't feel like I've tested it well.

**Comment by mdejean** [comment]
> Example before: Screenshot from 2021-11-22 11-45-16 after: Screenshot from 2021-11-22 11-45-29

Any idea why the bottom right lane lost its right turn?

File: map_model/src/objects/turn.rs (hunk 2 above)
Comment (dabreegster) [reply map_model/src/objects/turn.rs]: Thanks for the explanation.

> I'd like to think about not storing turn geometry at all and replacing the intersection in conflicts_with to something like a table lookup

IIRC, prior profiling showed that `conflicts_with` can be a hot-spot. A table lookup could have performance benefits as well. I agree we shouldn't tackle that in this PR though.

I'd vote for the conflict table to be built lazily, probably per intersection. If we tried to do it all upfront after loading a map, it'll be slow on the web, since we don't have multithreading working there yet. #786 is related -- we can calculate everything about turns as needed from other map structures, but it makes web slower

**Comment by dabreegster** [comment]
Just a heads up, I'm in the process of packing/moving and might be slow looking at PRs for a few days. :(

**Review by xorozo — APPROVED** [review]

**Comment by mdejean** [comment]
I thought about this for vehicle turns, but render/pedestrian also uses the turn angle of pedestrian turns. Right now this looks like pedestrians always wanting to turn around while waiting for a crosswalk:
image

**Comment by dabreegster** [comment]
I could see this as 2 independent choices:
- should the agent/on-the-road turn arrows try to use exact angles, or be binned into a few categories (left, straight, right)? How much should we communicate at the 5-way intersections with "slight right" and "sharp right"?
- what's the overall angle of a turn -- the straight line's angle from the first to the last point?

The waiting for a crosswalk bug looks like it's pointing to the first angle in the crosswalk's polyline, right? Or if the pedestrian was northbound, is this showing up as a right turn?

For pedestrians, I feel like just pointing directly from their current position to the end of the crosswalk they're waiting on would be most clear. Thoughts?

**Comment by mdejean** [comment]
> The waiting for a crosswalk bug looks like it's pointing to the first angle in the crosswalk's polyline, right? Or if the pedestrian was northbound, is this showing up as a right turn?

It's sort of pointing as though the crosswalk was a U turn, but also backwards because it's a ContraflowTurn.

> For pedestrians, I feel like just pointing directly from their current position to the end of the crosswalk they're waiting on would be most clear. Thoughts?

That was the case before but now brings the issue that with ContraflowTurns we'd have to figure out which is the the other end of the crosswalk. I've dealt with this by just pointing them at the middle - close enough!

**Comment by dabreegster** [comment]
>  I've dealt with this by just pointing them at the middle - close enough!

I mean, this is how I behave as a pedestrian in confusing situations. ;)  I think this solution makes sense. Let me know when I should review the code in depth

**Comment by mdejean** [comment]
It's ready. #820 covered a lot of the intersections where I was worried about the new turn geometry.

**Comment by mdejean** [comment]
> should the agent/on-the-road turn arrows try to use exact angles, or be binned into a few categories (left, straight, right)? How much should we communicate at the 5-way intersections with "slight right" and "sharp right"?

Certainly a possibility, and it would help stop arrows from bunching up (e.g. tempe:), but really necessary?

image

**Comment by dabreegster** [comment]
Code looks fine, minus some differences with the version of `lyon_geom` that got picked up... my `Cargo.lock` (which is under version control) is using `0.16.2`, so I had to make a few changes to match the slightly older API. We could upgrade, but we should do it separate from this PR and with caution -- a few times in the past, lyon upgrades have broken SVG text rendering.

I've regenerated everything and I'm looking at the diffs now...

- Map file size increases with the more detailed turn geometry. West Seattle gains about 1MB, as an example. This is fine, but something to keep in mind. #786 would let us save all the space spent serializing turns, but it requires figuring out web workers first to not degrade load times on the web.
- Agent turn arrows all look intuitive enough to me.

Before:
Screenshot from 2021-12-23 16-44-42
After -- much nicer!
Screenshot from 2021-12-23 16-44-53

Some turns got much curvier, looks way more realistic. Before in Tempe:
Screenshot from 2021-12-23 16-48-39
After:
Screenshot from 2021-12-23 16-48-49

This left turn from the cyclepath in Tempe is particularly impressive:
Screenshot from 2021-12-23 16-50-14

Looks great! Thanks again for this massive PR, I'll upload the new data and merge.
